### PR TITLE
PE schema adjustments

### PR DESCRIPTION
**Pull request by segraef**
Covering

- https://github.com/Azure/Azure-Verified-Modules/issues/620
- https://github.com/Azure/bicep-registry-modules/issues/946
- https://github.com/Azure/bicep-registry-modules/issues/1042

### CODE REVIEW

**Comment by segraef** [comment]
Covering

- https://github.com/Azure/Azure-Verified-Modules/issues/620
- https://github.com/Azure/bicep-registry-modules/issues/946

File: docs/static/includes/interfaces/int.pe.udt.schema1.bicep [hunk 1]
```diff
@@ -95,7 +95,7 @@ module <singularMainResourceType>_privateEndpoints 'br/public:avm/res/network/pr
     privateDnsZoneResourceIds: privateEndpoint.?privateDnsZoneResourceIds
     roleAssignments: privateEndpoint.?roleAssignments
     tags: privateEndpoint.?tags ?? tags
-    manualPrivateLinkServiceConnections: privateEndpoint.?manualPrivateLinkServiceConnections
+    manualPrivateLinkServiceConnections: !empty(privateEndpoint.?privateLinkServiceConnections) ? [] : privateEndpoint.?manualPrivateLinkServiceConnections
```
Comment (AlexanderSehr) [inline docs/static/includes/interfaces/int.pe.udt.schema1.bicep:98]:
```suggestion
    manualPrivateLinkServiceConnections: privateEndpoint.?privateLinkServiceConnections ?? []
```

File: docs/static/includes/interfaces/int.pe.udt.schema2.bicep [hunk 2]
```diff
@@ -95,7 +95,7 @@ module <singularMainResourceType>_privateEndpoints 'br/public:avm/res/network/pr
     privateDnsZoneResourceIds: privateEndpoint.?privateDnsZoneResourceIds
     roleAssignments: privateEndpoint.?roleAssignments
     tags: privateEndpoint.?tags ?? tags
-    manualPrivateLinkServiceConnections: privateEndpoint.?manualPrivateLinkServiceConnections
+    manualPrivateLinkServiceConnections: !empty(privateEndpoint.?privateLinkServiceConnections) ? [] : privateEndpoint.?manualPrivateLinkServiceConnections
```
Comment (AlexanderSehr) [inline docs/static/includes/interfaces/int.pe.udt.schema2.bicep:98]:
```suggestion
    manualPrivateLinkServiceConnections: privateEndpoint.?manualPrivateLinkServiceConnections ?? []
```

File: docs/static/includes/interfaces/int.pe.udt.schema1.bicep (hunk 1 above)
Comment (segraef) [reply docs/static/includes/interfaces/int.pe.udt.schema1.bicep:98]: Don't you mean `manualPrivateLinkServiceConnections: privateEndpoint.?privateLinkServiceConnections ?? []`?
I want `manualPrivateLinkServiceConnections` to be used only when `privateLinkServiceConnections` is `[]`.

File: docs/static/includes/interfaces/int.pe.udt.schema2.bicep (hunk 2 above)
Comment (segraef) [reply docs/static/includes/interfaces/int.pe.udt.schema2.bicep:98]: Don't you mean `manualPrivateLinkServiceConnections: privateEndpoint.?privateLinkServiceConnections ?? []`?
I want `manualPrivateLinkServiceConnections` to be used only when `privateLinkServiceConnections` is `[]`.

Comment (segraef) [reply docs/static/includes/interfaces/int.pe.udt.schema2.bicep:98]: Anyway, adding pre-constructed payload as per https://github.com/Azure/bicep-registry-modules/issues/946.

File: docs/static/includes/interfaces/int.pe.udt.schema1.bicep (hunk 1 above)
Comment (segraef) [reply docs/static/includes/interfaces/int.pe.udt.schema1.bicep:98]: Anyway, adding pre-constructed payload as per https://github.com/Azure/bicep-registry-modules/issues/946.

Comment (AlexanderSehr) [reply docs/static/includes/interfaces/int.pe.udt.schema1.bicep]: You're right, it's not correct. It should've been

```bicep
manualPrivateLinkServiceConnections: privateEndpoint.?manualPrivateLinkServiceConnections ?? []
```

File: docs/static/includes/interfaces/int.pe.udt.schema2.bicep (hunk 2 above)
Comment (AlexanderSehr) [reply docs/static/includes/interfaces/int.pe.udt.schema2.bicep]: As above, this should've been

```bicep
manualPrivateLinkServiceConnections: privateEndpoint.?manualPrivateLinkServiceConnections ?? []
```

**Comment by AlexanderSehr** [comment]
cc: @ChrisSidebotham, @jtracey93

**Comment by AlexanderSehr** [comment]
cc: @ChrisSidebotham, @jtracey93 :D

File: docs/static/includes/interfaces/int.pe.udt.schema1.bicep [hunk 3]
```diff
@@ -75,7 +75,7 @@ module <singularMainResourceType>_privateEndpoints 'br/public:avm/res/network/pr
   name: '${uniqueString(deployment().name, location)}-<singularMainResourceType>-PrivateEndpoint-${index}'
   params: {
     // Variant 1: A default service can be assumed (i.e., for services that only have one private endpoint type)
-    privateLinkServiceConnections: [
+    privateLinkServiceConnections: (!empty(privateEndpoint.?manualPrivateLinkServiceConnections) ? [] : [
```
Comment (AlexanderSehr) [inline docs/static/includes/interfaces/int.pe.udt.schema1.bicep:78]: Please note, this currently does not cover the suggestion of https://github.com/Azure/bicep-registry-modules/issues/946

**Review by AlexanderSehr — CHANGES_REQUESTED** [review]
All in all, I'd suggest we apply the combined suggestions @tyconsulting provided. That would mean
- Update the input schema to introduce an optional/nullable parameter `isManualConnection`
- Add it's usage to the PE deployment including the updated field for the connection name.

Example for usage schema 1:
```Bicep
module <singularMainResourceType>_privateEndpoints 'br/public:avm/res/network/private-endpoint:X.Y.Z' = [for (privateEndpoint, index) in (privateEndpoints ?? []): {
  name: '${uniqueString(deployment().name, location)}-<singularMainResourceType>-PrivateEndpoint-${index}'
  params: {
    name: privateEndpoint.?name ?? 'pep-${last(split(<singularMainResourceType>.id, '/'))}-${privateEndpoint.?service ?? <defaultServiceName>}-${index}'
    privateLinkServiceConnections: privateEndpoint.?isManualConnection != true ? [
      {
        name: privateEndpoint.?privateLinkServiceConnectionName ?? '${last(split(<singularMainResourceType>.id, '/'))}-${privateEndpoint.?service ?? '<defaultServiceName>'}-${index}'
        properties: {
          privateLinkServiceId: <singularMainResourceType>.id
          groupIds: [
            privateEndpoint.?service ?? <defaultServiceName>
          ]
        }
      }
    ] : null
    manualPrivateLinkServiceConnections: privateEndpoint.?isManualConnection == true ? [
      {
        name: privateEndpoint.?privateLinkServiceConnectionName ?? '${last(split(<singularMainResourceType>.id, '/'))}-${privateEndpoint.?service ?? '<defaultServiceName>'}-${index}'
        properties: {
          privateLinkServiceId: workspace.id
          groupIds: [
            privateEndpoint.?service ?? <defaultServiceName>
          ]
          requestMessage: privateEndpoint.?manualConnectionRequestMessage ?? 'Manual approval required.'
        }
      }
    ] : null
    subnetResourceId: privateEndpoint.subnetResourceId
    enableTelemetry: privateEndpoint.?enableTelemetry ?? enableTelemetry
    location: privateEndpoint.?location ?? reference(split(privateEndpoint.subnetResourceId, '/subnets/')[0], '2020-06-01', 'Full').location
    lock: privateEndpoint.?lock ?? lock
    privateDnsZoneGroupName: privateEndpoint.?privateDnsZoneGroupName
    privateDnsZoneResourceIds: privateEndpoint.?privateDnsZoneResourceIds
    roleAssignments: privateEndpoint.?roleAssignments
    tags: privateEndpoint.?tags ?? tags
    customDnsConfigs: privateEndpoint.?customDnsConfigs
    ipConfigurations: privateEndpoint.?ipConfigurations
    applicationSecurityGroupResourceIds: privateEndpoint.?applicationSecurityGroupResourceIds
    customNetworkInterfaceName: privateEndpoint.?customNetworkInterfaceName
  }
}]
```

**Comment by segraef** [comment]
Okay I tried to find a more elegant way to avoid introducing an additional parameter such as `isManualConnection` to validate if it's a manual connection. Correct me if I'm wrong but it's always either `privateLinkServiceConnectionName` or `manualPrivateLinkServiceConnections`, never both so I can go the way of just validate if the array manualPrivateLinkServiceConnections is filled or not.

**Comment by AlexanderSehr** [comment]
> Okay I tried to find a more elegant way to avoid introducing an additional parameter such as `isManualConnection` to validate if it's a manual connection. Correct me if I'm wrong but it's always either `privateLinkServiceConnectionName` or `manualPrivateLinkServiceConnections`, never both so I can go the way of just validate if the array manualPrivateLinkServiceConnections is filled or not. See [8d4ae89](https://github.com/Azure/Azure-Verified-Modules/commit/8d4ae897dd556ff405d2e28b5dd2ec8866b82ddb) .

Unfortunately, the suggestion won't work if I understand it correctly. My main question would be how the parameter `privateEndpoint.manualPrivateLinkServiceConnections` would look like?

It's value is not used. And the reason it isn't is because there can always only be either the automatic - or - manual connection, and in either case only one (i.e., it's never a list of e.g. multiple manual service connections for one private endpoint). I would suspect that's why the boolean parameter was suggested instead (not in addition). If the value is true, the single manual service connection is deployed and the automatic one is not. If the value is not true (i.e., is null) the automatic connection is deployed, but the manual not.

**Comment by segraef** [comment]
testing [sa](https://github.com/segraef/bicep-registry-modules/actions/runs/8002637833) via https://github.com/segraef/bicep-registry-modules/pull/7

File: docs/static/includes/interfaces/int.pe.udt.schema1.bicep [hunk 4]
```diff
@@ -75,18 +78,30 @@ module <singularMainResourceType>_privateEndpoints 'br/public:avm/res/network/pr
...
+        }
+      }
+    ] : null
+    manualPrivateLinkServiceConnections: privateEndpoint.?manualPrivateLinkServiceConnections == true ? [
+      {
+        name: privateEndpoint.?privateLinkServiceConnectionName ?? '${last(split(<singularMainResourceType>.id, '/'))}-${privateEndpoint.?service ?? '<defaultServiceName>'}-${index}'
+        properties: {
+          privateLinkServiceId: workspace.id
+          groupIds: [
+            privateEndpoint.?service ?? <defaultServiceName>
```
Comment (AlexanderSehr) [inline docs/static/includes/interfaces/int.pe.udt.schema1.bicep:99]:
```suggestion
            privateEndpoint.?service ?? '<defaultServiceName>'
```

File: docs/static/includes/interfaces/int.pe.udt.schema1.bicep [hunk 5]
```diff
@@ -19,6 +19,12 @@ type privateEndpointType = {
   @description('Optional. The private DNS zone groups to associate the private endpoint with. A DNS zone group can support up to 5 DNS zones.')
   privateDnsZoneResourceIds: string[]?
 
+  @description('Optional. Manual PrivateLink Service Connections.')
```
Comment (AlexanderSehr) [inline docs/static/includes/interfaces/int.pe.udt.schema1.bicep:22]:
```suggestion
  @description('Optional. If Manual Private Link Connection is required.')
```

File: docs/static/includes/interfaces/int.pe.udt.schema2.bicep [hunk 6]
```diff
@@ -19,6 +19,12 @@ type privateEndpointType = {
   @description('Optional. The private DNS zone groups to associate the private endpoint with. A DNS zone group can support up to 5 DNS zones.')
   privateDnsZoneResourceIds: string[]?
 
+  @description('Optional. Manual PrivateLink Service Connections.')
```
Comment (AlexanderSehr) [inline docs/static/includes/interfaces/int.pe.udt.schema2.bicep:22]:
```suggestion
  @description('Optional. If Manual Private Link Connection is required.')
```

File: docs/static/includes/interfaces/int.pe.udt.schema2.bicep [hunk 7]
```diff
@@ -19,6 +19,12 @@ type privateEndpointType = {
...
+  isManualConnection: bool?
+
+  @description('Optional. The name of the request message for manual approval.')
```
Comment (AlexanderSehr) [inline docs/static/includes/interfaces/int.pe.udt.schema2.bicep:25]:
```suggestion
  @description('Optional. A message passed to the owner of the remote resource with the manual connection request. Restricted to 140 chars.')
```

File: docs/static/includes/interfaces/int.pe.udt.schema1.bicep [hunk 8]
```diff
@@ -19,6 +19,12 @@ type privateEndpointType = {
...
+  isManualConnection: bool?
+
+  @description('Optional. The name of the request message for manual approval.')
```
Comment (AlexanderSehr) [inline docs/static/includes/interfaces/int.pe.udt.schema1.bicep:25]:
```suggestion
  @description('Optional. A message passed to the owner of the remote resource with the manual connection request. Restricted to 140 chars.')
```

File: docs/static/includes/interfaces/int.pe.udt.schema2.bicep [hunk 9]
```diff
@@ -74,19 +77,31 @@ param privateEndpoints privateEndpointType
 module <singularMainResourceType>_privateEndpoints 'br/public:avm/res/network/private-endpoint:X.Y.Z' = [for (privateEndpoint, index) in (privateEndpoints ?? []): {
   name: '${uniqueString(deployment().name, location)}-<singularMainResourceType>-PrivateEndpoint-${index}'
   params: {
-    // Variant 2: A default service cannot be assumed (i.e., for services that have more than one private endpoint type, like Storage Account)
-    privateLinkServiceConnections: [
+    // Variant 1: A default service can be assumed (i.e., for services that only have one private endpoint type)
```
Comment (AlexanderSehr) [inline docs/static/includes/interfaces/int.pe.udt.schema2.bicep:80]:
```suggestion
    // Variant 2: A default service cannot be assumed (i.e., for services that have more than one private endpoint type, like Storage Account)
```

**Comment by segraef** [comment]
[max](https://github.com/segraef/bicep-registry-modules/actions/runs/8002637833/job/21856426030) test succeeded with concurrent PE deployments for SA:

image

testing other multi-PE services as per https://learn.microsoft.com/en-us/azure/private-link/private-endpoint-overview#private-link-resource

Private-link resource name | Resource type | Sub-resources
-- | -- | --
Azure Automation | Microsoft.Automation/automationAccounts | Webhook, DSCAndHybridWorker
Azure Backup | Microsoft.RecoveryServices/vaults | AzureBackup, AzureSiteRecovery
Azure Batch | Microsoft.Batch/batchAccounts | batchAccount, nodeManagement
Azure Cosmos DB | Microsoft.AzureCosmosDB/databaseAccounts | SQL, MongoDB, Cassandra, Gremlin, Table
Azure Databricks | Microsoft.Databricks/workspaces | databricks_ui_api, browser_authentication
Azure Media Services | Microsoft.Media/mediaservices | keydelivery, liveevent, streamingendpoint
Azure Storage | Microsoft.Storage/storageAccounts | Blob (blob, blob_secondary)Table (table, table_secondary)Queue (queue, queue_secondary)File (file, file_secondary)Web (web, web_secondary)Dfs (dfs, dfs_secondary)
Azure Synapse | Microsoft.Synapse/privateLinkHubs | web
Azure Synapse Analytics | Microsoft.Synapse/workspaces | Sql, SqlOnDemand, Dev

**Review by AlexanderSehr — APPROVED** [review]

**Pull request merged**